File: server/src/protocol/ErrorResponse.cpp
#include "ErrorResponse.h"
#include "ResponseVisitor.h"

using namespace std;
using namespace communication;

ErrorResponse::ErrorResponse(ErrorType type)
        : Response(10, ResponseType::ERROR),
          type_(type)
{}

ErrorResponse::~ErrorResponse()
{}

vector<uint8_t> ErrorResponse::getFrameBytes()
{
    initializeDataSize();

    vector<uint8_t > frame = getHeader();
    frame.push_back(static_cast<uint8_t>(responseType_));
    frame.push_back(static_cast<uint8_t>(type_));

    return frame;
}

string ErrorResponse::getName()
{
    return string("ErrorResponse");
}

void ErrorResponse::accept(ResponseVisitor &visitor)
{
    visitor.visit(*this);
}

ErrorType ErrorResponse::getErrorType() const
{
    return type_;
}

void ErrorResponse::setErrorType(ErrorType type)
{
    type_ = type;
}

void ErrorResponse::initializeDataSize()
{
    uint16_t dataSize = sizeof(responseType_);
    dataSize += sizeof(type_);

    setDataSize(dataSize);
}

File: mobile_client/app/src/main/java/wpam/mobile_client/protocol/ErrorResponse.java
package wpam.mobile_client.protocol;

import java.util.ArrayList;

public final class ErrorResponse extends Response
	{
		private ErrorType type_;

		public ErrorResponse(ErrorType type)
		{
			super(10, ResponseType.ERROR);
			this.type_ = type;
		}

		public final void close()
		{
			super.close();
		}

		@Override
		public ArrayList<Integer> getFrameBytes()
		{
			initializeDataSize();

			ArrayList<Integer> frame = getHeader();
			frame.add(responseType_.getValue());
			frame.add(type_.getValue());

			return frame;
		}
		@Override
		public String getName()
		{
			return new String("ErrorResponse");
		}

		@Override
		public void accept(ResponseVisitor visitor)
		{
			visitor.visit(this);
		}

		public ErrorType getErrorType()
		{
			return this.type_;
		}

		public void setErrorType(ErrorType type)
		{
			this.type_ = type;
		}

		@Override
		protected void initializeDataSize()
		{
			setDataSize(2);
		}
	}